### PR TITLE
Fix parsing SemanticVersion build label from version string

### PR DESCRIPTION
**Pull request by JoeRobich**
# PR Summary

Improve the SemanticVersion parsing logic to handle missing cases.

## PR Context

Parsing semantic version strings of the form `major.minor.patch+release` is broken.

Here is an example from PowerShell 7.2:
```
~/Source> $v1 = [System.Management.Automation.SemanticVersion]::new("1.0.0+META")
MethodInvocationException: Exception calling ".ctor" with "1" argument(s): "Input string was not in a correct format."
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/engine/Basic/SemanticVersion.Tests.ps1 [hunk 1]
```diff
@@ -22,6 +22,14 @@ Describe "SemanticVersion api tests" -Tags 'CI' {
             $v.BuildLabel | Should -BeNullOrEmpty
             $v.ToString() | Should -Be "1.0.0"
 
+            $v = [SemanticVersion]::new("1.2.3+BLD.a1-xxx.03")
+            $v.Major | Should -Be 1
+            $v.Minor | Should -Be 0
+            $v.Patch | Should -Be 0
```
Comment (iSazonov) [inline test/powershell/engine/Basic/SemanticVersion.Tests.ps1:28]:
```suggestion
            $v.Major | Should -Be 1
            $v.Minor | Should -Be 2
            $v.Patch | Should -Be 3
```

**Comment by JoeRobich** [comment]
@iSazonov I believe this is ready for a review. I choose a good test case because my initial fix wasn't quite right according to the spec. Now I believe we are properly parsing the labels.

**Review by iSazonov — APPROVED** [review]

File: test/powershell/engine/Basic/SemanticVersion.Tests.ps1 [hunk 2]
```diff
@@ -22,6 +22,14 @@ Describe "SemanticVersion api tests" -Tags 'CI' {
...
             $v.ToString() | Should -Be "1.0.0"
 
+            $v = [SemanticVersion]::new("1.2.3+BLD.a1-xxx.03")
+            $v.Major | Should -Be 1
+            $v.Minor | Should -Be 2
+            $v.Patch | Should -Be 3
+            $v.PreReleaseLabel | Should -BeNullOrEmpty
+            $v.BuildLabel | Should -Be "BLD.a1-xxx.03"
+            $v.ToString() | Should -Be "1.2.3+BLD.a1-xxx.03"
+
```
Comment (daxian-dbw) [inline test/powershell/engine/Basic/SemanticVersion.Tests.ps1:32]: Can you add a test case to cover `1.0.0+META`? That's what you fixed in this PR but there is no test for it.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: test/powershell/engine/Basic/SemanticVersion.Tests.ps1 (hunk 2 above)
Comment (JoeRobich) [reply test/powershell/engine/Basic/SemanticVersion.Tests.ps1:32]: Oh sure. Anything after the '+' is considered metadata (or META) which is why I thought this test case was appropriate.

File: test/powershell/engine/Basic/SemanticVersion.Tests.ps1 [hunk 3]
```diff
@@ -22,6 +22,14 @@ Describe "SemanticVersion api tests" -Tags 'CI' {
             $v.BuildLabel | Should -BeNullOrEmpty
             $v.ToString() | Should -Be "1.0.0"
 
+            $v = [SemanticVersion]::new("1.2.3+BLD.a1-xxx.03")
+            $v.Major | Should -Be 1
+            $v.Minor | Should -Be 2
+            $v.Patch | Should -Be 3
+            $v.PreReleaseLabel | Should -BeNullOrEmpty
+            $v.BuildLabel | Should -Be "BLD.a1-xxx.03"
+            $v.ToString() | Should -Be "1.2.3+BLD.a1-xxx.03"
```
Comment (JoeRobich) [inline test/powershell/engine/Basic/SemanticVersion.Tests.ps1:31]:
```suggestion
            $v = [SemanticVersion]::new("1.2.3+BLD.a1-xxx.03")
            $v.Major | Should -Be 1
            $v.Minor | Should -Be 2
            $v.Patch | Should -Be 3
            $v.PreReleaseLabel | Should -BeNullOrEmpty
            $v.BuildLabel | Should -Be "BLD.a1-xxx.03"
            $v.ToString() | Should -Be "1.2.3+BLD.a1-xxx.03"

            $v = [SemanticVersion]::new("1.0.0+META")
            $v.Major | Should -Be 1
            $v.Minor | Should -Be 0
            $v.Patch | Should -Be 0
            $v.PreReleaseLabel | Should -BeNullOrEmpty
            $v.BuildLabel | Should -Be "META"
            $v.ToString() | Should -Be "1.0.0+META"
```

File: test/powershell/engine/Basic/SemanticVersion.Tests.ps1 (hunk 2 above)
Comment (JoeRobich) [reply test/powershell/engine/Basic/SemanticVersion.Tests.ps1:32]: But now that I look back over the code. Yes, I wasn't testing the change. Thanks!

**Review by daxian-dbw — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@JoeRobich Thanks for your contribution!

**Pull request merged**